### ISSUE TITLE
using $this in pure function

### ISSUE THREAD

**Issue by ptomulik** [vimeo/psalm#7438]
I'd like discuss  psalm's concept of functions' purity. The [article](https://psalm.dev/articles/immutability-and-beyond) introducing "@psalm-pure" refers to [wikipedia's definition of pure function](https://en.wikipedia.org/wiki/Pure_function), which states:

> A pure function is a function that has the following properties:[1][2]
>  1. the function return values are identical for identical arguments (...), and
>  2. the function application has no side effects (...).

Then an [ImpureVariable error example](https://psalm.dev/docs/running_psalm/issues/ImpureVariable/) appears, which demonstrates how psalm names a pure function impure.

I can't understand why ``$this`` is so exceptional in this context? Shouldn't it be treated as a regular parameter to an instance method? I think it should, and I'm not alone here, for example: https://stackoverflow.com/a/50466413/2186051. In my opinion, the difference between regular parameter ``$obj`` and ``$this`` is just the notation. It doesn't differ much semantically, though. Here are some facts:

- ``$this`` is not a global variable from method's perspective, it's not either static variable/property; if it was, then the following code should not work as it currently works:
```php
<?php declare(strict_types=1);
final class C {
  public function f(): void
  {
    $s = new self();
    $s->g();
    // if $this was global/static then we had $this === $s at this point,
    // but we have $this === $c
  }
  public function g(): void
  {
  }
}
$c = new C();
$c->f();
```
- ``$this`` is not instantiated (randomly) inside a method, it's always provided by caller.

So, ``$this`` has all the symptoms of function's parameter, except the notation (it's not passed via argument list within parentheses and does not appear explicitly on parameter list in method's signature). But this is only how we write code, the program still works as if ``$this`` was a regular parameter.

There are some restrictions regarding ``$this`` that don't apply to regular parameters/variables:

- ``$this`` can't be (re)assigned,
- ``$this`` can't be used as a name for explicit parameter in function signature,

but this doesn't matter when discussing methods' purity.

The concept of "instance variable" is similar across many languages. Python makes things more explicit - each instance method declares explicitly a parameter called ``self`` on its parameter list. The notation for calling instance method remains same - ``obj.method()``:

```python
class C:  
    def foo(self):  
        return self

c = C()
c.foo()
```
I think ``$this`` should be used by psalm as a regular parameter, and documentation should state this. The purity analysis works quite well for object references provided via parameters list and, perhaps, the same code could be used directly for ``$this``. This could reduce code size and cut the unnecessary discussions of how ``$this`` is different from the same instance but provided by parameter
declared within parentjeses.

**Comment by AndrolGenhald** [comment]
Yes, the major difference between `$this` and other arguments in regards to purity is that it's effectively always passed by reference, and we do indeed handle references correctly in this case, allowing them but preventing assignment: https://psalm.dev/r/0215ecf4aa, https://psalm.dev/r/677d9fb1ad (although having a reference argument is pointless if you're not able to modify it, so maybe we should raise an issue there, except in the case of overriding a method and marking it as pure but the signature has to match the parent that takes a reference).

I agree [this case](https://psalm.dev/r/96fef23e6d) should be handled the same as the one above, but [this](https://psalm.dev/r/4f3aa154ee) should still be disallowed.

So basically, `@psalm-mutation-free` should be deprecated and aliased to `@psalm-pure`: https://psalm.dev/r/b541ebb36e, https://psalm.dev/r/4753d09ec0.

Perhaps @muglug has some thoughts on the matter since he implemented it?

**Comment by ptomulik** [comment]
> Yes, the major difference between `$this` and other arguments in regards to purity is that it's effectively always passed by reference, and we do indeed handle references correctly in this case, allowing them but preventing assignment: https://psalm.dev/r/0215ecf4aa, https://psalm.dev/r/677d9fb1ad (although having a reference argument is pointless if you're not able to modify it, so maybe we should raise an issue there, except in the case of overriding a method and marking it as pure but the signature has to match the parent that takes a reference).

Well, in some languages the first example is considered to be an impure function. Pure function's result can't vary with a value assigned to mutable reference (pointer in C). This is actually mentioned in [wikipedia](https://en.wikipedia.org/wiki/Pure_function) and the page gives this example of impure function in C:
```C
int f(int* x) {
  return *x;
}
```
This is, because in several languages a reference is considered as a value itself (a number representing memory address for example). In PHP references are just aliases to symbol table (different names for a variable), so scalar references are indistinguishable from scalar parameters in this context - as long as we don't modify the content under reference. Things complicate when we consider objects references.

> 
> I agree [this case](https://psalm.dev/r/96fef23e6d) should be handled the same as the one above, but [this](https://psalm.dev/r/4f3aa154ee) should still be disallowed.

Not exactly, in [your example](https://psalm.dev/r/96fef23e6d) the function should be considered impure, because its result varies with a value behind mutable reference (the value of mutable object's property ``$bar``). Providing same object twice may result with different results because object's state can change in the meantime (outside of the function). Let me simplify the example to [this](https://psalm.dev/r/11d384985b). It's indeed impure. A pure function can only access values behind immutable references (immutable objects' properties), so impure function in Psalm shall only work with objects tagged as ``@psalm-impure``: https://psalm.dev/r/39d3ff0bc7.

Referring object reference parameters (as opposed to object's compounds) should be handled same way as refering regular parameters. And it works in psalm for regular parameters, but not for ``$this``: https://psalm.dev/r/7d6dae4339

Now, my postulate is, that pure instance methods should handle ``$this`` same way as regular parameters are handled. It's not the case, however: https://psalm.dev/r/7d6dae4339

> So basically, `@psalm-mutation-free` should be deprecated and aliased to `@psalm-pure`: https://psalm.dev/r/b541ebb36e, https://psalm.dev/r/4753d09ec0.

Don't think so. Pure and mutation-free are not same. Any pure function must be mutation free and (additionally) its result can't depend on side state (it can only vary with arguments). So in short: @psalm-mutation-free => the function does not modify side state, @psalm-pure => @psalm-mutation-free + the result of function does not depend on side state.

> 
> Perhaps @muglug has some thoughts on the matter since he implemented it?

**Comment by AndrolGenhald** [comment]
> ... Let me simplify the example to [this](https://psalm.dev/r/11d384985b). It's indeed impure.

Oh, I would not have expected that. I was thinking it would still be pure because the return value is identical for identical inputs, but I suppose here "identical" means same object even if properties are changed (eg treating an object like a pointer)? This seems to be the main point of difference between my idea of how it should work and yours.

> This is actually mentioned in [wikipedia](https://en.wikipedia.org/wiki/Pure_function) and the page gives this example of impure function in C:
> 
> ```c
> int f(int* x) {
>   return *x;
> }
> ```

I can understand this for C where the value of the pointer is can be the same while the referred to value is different, but in PHP references are transparent, you can't even tell that a particular variable is a reference at runtime. It's impossible for this function to know that both calls use the same variable: https://3v4l.org/dH2sF (even in the case where the variable passed is itself a reference https://3v4l.org/79CbZ)
It seems to me that a reference argument that's never assigned to is effectively indistinguishable from a non-reference argument.

The pointer example from Wikipedia does explain why using object properties would be different though, since an object is effectively a reference that you _can_ compare by pointer value rather than by referenced value: https://3v4l.org/treno

**Comment by ptomulik** [comment]
> > ... Let me simplify the example to [this](https://psalm.dev/r/11d384985b). It's indeed impure.
> 
> Oh, I would not have expected that. I was thinking it would still be pure because the return value is identical for identical inputs, but I suppose here "identical" means same object even if properties are changed (eg treating an object like a pointer)? This seems to be the main point of difference between my idea of how it should work and yours.

That's the point! Philosophically we could consent, that two objects are identical if and only if they're of excactly same type and have exactly same property values (comparison by value). But then we would had a hard time explaining, why we didn't account the PHP's [===](https://www.php.net/manual/en/language.oop5.object-comparison.php) operator in this philosophical concept.


> > This is actually mentioned in [wikipedia](https://en.wikipedia.org/wiki/Pure_function) and the page gives this example of impure function in C:
> > ```c
> > int f(int* x) {
> >   return *x;
> > }
> > ```
> 
> I can understand this for C where the value of the pointer is can be the same while the referred to value is different, but in PHP references are transparent, you can't even tell that a particular variable is a reference at runtime. It's impossible for this function to know that both calls use the same variable: https://3v4l.org/dH2sF (even in the case where the variable passed is itself a reference https://3v4l.org/79CbZ) It seems to me that a reference argument that's never assigned to is effectively indistinguishable from a non-reference argument.
> 
> The pointer example from Wikipedia does explain why using object properties would be different though, since an object is effectively a reference that you _can_ compare by pointer value rather than by referenced value: https://3v4l.org/treno

Exactly.

**Comment by ptomulik** [comment]
Anyway, making this long story short: objects are handled properly by psalm, except for ``$this``. This should be fixed. :)

**Comment by marcosh** [comment]
I think that having the possibility to say that the output of a function/method depends ONLY on its explicit inputs has its value. Sometime it is exactly what you want to say. 

On the other hand, some other times you would like to treat `$this` as a method argument and relax your guarantees to compute identical values on the same inputs only when you are dealing with the same instance.

That's why I would consider wise to have both concepts at hand, with the possibility to choose every time what you want.

I tried to sketch some notes on this topic [here](https://hackmd.io/hkIJIhA_STCHxmFnxs9IsQ)

**Comment by AndrolGenhald** [comment]
@marcosh That's a good start, but I think it would be nice to add explicit notes about when accessing external state is allowed. Should comparing to a global be allowed in `@psalm-mutation-free`? It's name would imply yes, but Psalm currently assumes that it's not allowed in some places (the implementation is inconsistent).

**Comment by marcosh** [comment]
@AndrolGenhald thanks for the hint. I updated the document with my view on the issue.

I think that no one of the annotations described should allow accessing global state. All of them provide some flavour of purity and accessing global state would definitely break purity.

**Comment by muglug** [comment]
Yeah, I think `@psalm-mutation-free` should match that document's definition of `@psalm-instance-pure` — one of the advantages of `@psalm-mutation-free` calls is they can be memoised in the typechecker, and that's unsound if they're comparing to global state.

**Comment by muglug** [comment]
Oh actually thinking about this again I think the correction that should be made is that `@psalm-immutable` objects must have `@psalm-instance-pure` methods, but we can keep the current definition of `@psalm-mutation-free` to mean "behaves like a property fetch".

**Comment by marcosh** [comment]
@muglug currently `@psalm-mutation-free` method can access mutable state. Are you suggesting to change this behaviour?

**Comment by marcosh** [comment]
@muglug on immutable classes `@psalm-instance-pure` and `@psalm-mutation-free` should coincide since there is no mutable state

**Comment by AndrolGenhald** [comment]
I think a useful distinction to make is to have a tag to describe "This method must consider objects as equal by reference", where accessing properties and returning a new object is disallowed, and a different tag to describe "This method must consider objects as equal by property value" which could allow returning a new object and accessing argument and `$this` properties but it must be return the same value (comparing by property) for the same arguments (comparing by property).

**Comment by muglug** [comment]
@marcosh A lot of thoughts in my head, sorry, and my memory was hazy.

The current behaviour with immutable classes is sound: global state is not accessible from `@psalm-immutable` methods: https://psalm.dev/r/2fd942b02a

We also prohibit accessing global state from `@psalm-mutation-free` methods: https://psalm.dev/r/3572aece25

**Comment by marcosh** [comment]
> I think a useful distinction to make is to have a tag to describe "This method must consider objects as equal by reference", where accessing properties and returning a new object is disallowed, and a different tag to describe "This method must consider objects as equal by property value" which could allow returning a new object and accessing argument and `$this` properties but it must be return the same value (comparing by property) for the same arguments (comparing by property).

personally I don't see that much value in this and I like the fact that pure constructors are allowed in pure code (i.e. calling twice a pure constructor you get two _equal_ things, even though not identical). Considering strict equality for objects is, as far as I see it, practically never what you'd like to have (I'll be happy to see a counterexample, though)

**Comment by AndrolGenhald** [comment]
> global state is not accessible from `@psalm-immutable` methods

See #7425.

> Considering strict equality for objects is, as far as I see it, practically never what you'd like to have

It can be useful sometimes: https://psalm.dev/r/0b421abed5
I'm sure there are better examples. Given that purity in other languages prevents accessing mutable state behind a pointer, I don't like changing that meaning for Psalm. I do think adding another tag to allow that could be useful though, as long as everything is well documented.

**Comment by marcosh** [comment]
@AndrolGenhald one possible issue that arises if you provide the choice on which equality to use, strict or not, is that it would make sense to possibly specify a different type of equality for every argument of the function, `$this` included, and for the output. Being able to specify that, for example, that the first and the third argument should use strict equality, while the second should not becomes quite a burden on the API.
